### PR TITLE
[WebNN EP] TFLite backend only supports limit ranges for Clip

### PR DESCRIPTION
**Pull request by Honry**

### CODE REVIEW

**Comment by Honry** [comment]
@guschmue, @fs-eire, @fdwr, PTAL, thanks!

**Comment by fs-eire** [comment]
It looks to me that the change is taking considerations of implementation details on more priority than standard specification. Although WebNN is using TFLite, it does not necessarily mean that (technically) it will always use it on every environment, and even if so, it is still possible that a future version of TFLite may support arbitrary min/max attribute of Clip/Clamp.

It is hard to control the browser version and platforms users are using, so they may use a different version. This is why a standard is important and in my understanding the code should stick to the spec as much as possible, instead of the status of a particular underlying implementation.

There are a few PRs that are doing similar and I assume it's because in Chromium WebNN's underlying CPU engine is switched from XNNPACK to TFLite. I understand that there are technical reasons to do this in order to make it E2E working for WebNN. However, in long term, it would probably be better to introduce a way to allow user to know the gap between the spec and the implementation. It can be feature detection via some WebNN API or introducing WebNN version. A developer should have a better way to know whether a node can be supported in the current WebNN environment other than knowing how the detailed status of the underlying engine, which is supposed to be transparent.

**Comment by Honry** [comment]
> It looks to me that the change is taking considerations of implementation details on more priority than standard specification. Although WebNN is using TFLite, it does not necessarily mean that (technically) it will always use it on every environment, and even if so, it is still possible that a future version of TFLite may support arbitrary min/max attribute of Clip/Clamp.
> 
> It is hard to control the browser version and platforms users are using, so they may use a different version. This is why a standard is important and in my understanding the code should stick to the spec as much as possible, instead of the status of a particular underlying implementation.
> 
> There are a few PRs that are doing similar and I assume it's because in Chromium WebNN's underlying CPU engine is switched from XNNPACK to TFLite. I understand that there are technical reasons to do this in order to make it E2E working for WebNN. However, in long term, it would probably be better to introduce a way to allow user to know the gap between the spec and the implementation. It can be feature detection via some WebNN API or introducing WebNN version. A developer should have a better way to know whether a node can be supported in the current WebNN environment other than knowing how the detailed status of the underlying engine, which is supposed to be transparent.

Thanks @fs-eire, very good point! It's really painful for users at current stage, as the spec and the implementation of WebNN is still under very active evolution. @fujunwei is working on these unsupported constraints by emulating or some other methods to fill up the gaps.

We will finally close the gap, and this [table](https://github.com/microsoft/onnxruntime/blob/main/js/web/docs/webnn-operators.md) will keep up-to-date to maintain the latest status of op support status and constraints.

For long term, once everything is stable and the gap is small (e.g. WebNN passes the Origin Trial in Chromium), we can note users more change info along with Chrome version. WDYT?

cc/ @huningxin, @ibelem

**Comment by fs-eire** [comment]
> > It looks to me that the change is taking considerations of implementation details on more priority than standard specification. Although WebNN is using TFLite, it does not necessarily mean that (technically) it will always use it on every environment, and even if so, it is still possible that a future version of TFLite may support arbitrary min/max attribute of Clip/Clamp.
> > It is hard to control the browser version and platforms users are using, so they may use a different version. This is why a standard is important and in my understanding the code should stick to the spec as much as possible, instead of the status of a particular underlying implementation.
> > There are a few PRs that are doing similar and I assume it's because in Chromium WebNN's underlying CPU engine is switched from XNNPACK to TFLite. I understand that there are technical reasons to do this in order to make it E2E working for WebNN. However, in long term, it would probably be better to introduce a way to allow user to know the gap between the spec and the implementation. It can be feature detection via some WebNN API or introducing WebNN version. A developer should have a better way to know whether a node can be supported in the current WebNN environment other than knowing how the detailed status of the underlying engine, which is supposed to be transparent.
> 
> Thanks @fs-eire, very good point! It's really painful for users at current stage, as the spec and the implementation of WebNN is still under very active evolution. @fujunwei is working on these unsupported constraints by emulating or some other methods to fill up the gaps.
> 
> We will finally close the gap, and this [table](https://github.com/microsoft/onnxruntime/blob/main/js/web/docs/webnn-operators.md) will keep up-to-date to maintain the latest status of op support status and constraints.
> 
> For long term, once everything is stable and the gap is small (e.g. WebNN passes the Origin Trial in Chromium), we can note users more change info along with Chrome version. WDYT?
> 
> cc/ @huningxin, @ibelem

Thank you for the information.

Is there existing discussion or proposal about:
- APIs that allows users to query the implementation status
- versioning

I think the gap may always exist because everything is moving: spec will update and accept new operators (or maybe deprecate less used operators too), and implementation will upgrade too. So I am not optimistic on "everything is stable" and I think a standard way to manage the gap is more reasonable to me.

**Comment by Honry** [comment]
> Is there existing discussion or proposal about:
>
> - APIs that allows users to query the implementation status

This issue https://github.com/webmachinelearning/webnn/issues/463 is talking about exposing operators/types status for each backend.

> - versioning

I am not aware of discussion about versioning. @huningxin, do you know?

> I think the gap may always exist because everything is moving: spec will update and accept new operators (or maybe deprecate less used operators too), and implementation will upgrade too. So I am not optimistic on "everything is stable" and I think a standard way to manage the gap is more reasonable to me.

You are right, I mean we can mark more change info once the gap is smaller and the change is not so frequent.

We have a [WebNN Status](https://webmachinelearning.github.io/webnn-status/) page to maintain the op impl status for each backend with Chrome version info.

Maybe we can add Chrome and ORT-Web version info to this [table](https://github.com/microsoft/onnxruntime/blob/main/js/web/docs/webnn-operators.md) in future.

**Comment by Honry** [comment]
BTW, @fs-eire, do you know how does WebGPU EP manage versioning?

File: js/web/docs/webnn-operators.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ operators and the supported opset domain/versions in **WebNN EP** by ONNX Runtim
 | BatchNormalization | ai.onnx(7-8, 9-13, 14, 15+) | batchNormalization | ✗ | ✓ | Only supports 'training_mode' value is 0, one output |
 | Cast | ai.onnx(7-8, 9-12, 13-18, 19-20, 21+) | cast | ✗ | ✓ | |
 | Ceil | ai.onnx(7-12, 13+) | ceil | ✓ | ✓ | |
-| Clip | ai.onnx(7-10, 11, 12, 13+) | clamp | ✓ | ✓ | |
+| Clip | ai.onnx(7-10, 11, 12, 13+) | clamp | ✓ | ✓ | WebNN CPU backend only supports 3 specific ranges: [0.0, infinity], [-1.0, 1.0], [0.0, 6.0] |
```
Comment (huningxin) [inline js/web/docs/webnn-operators.md:22]: There is a Chromium issue to support other ranges in TFLite backend: https://issues.chromium.org/issues/326156496. You may want to mention it here.

**Comment by huningxin** [comment]
@fs-eire , thanks for your feeback!

> This is why a standard is important and in my understanding the code should stick to the spec as much as possible, instead of the status of a particular underlying implementation.

+1

> it is still possible that a future version of TFLite may support arbitrary min/max attribute of Clip/Clamp.

Before underlying runtime like TFLite gets that support, WebNN implementation can emulate by composition. In WebNN spec, for each operators that can be decomposed, there is an emulation sample code. For example, [`clamp`](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-clamp) can be emulated by `min` and `max` operators as following code sample

```js
if (options.minValue === undefined) {
  if (options.maxValue === undefined) {
    return input;
  } else {
    return builder.min(input, builder.constant(options.maxValue));
  }
} else {
  if (options.maxValue === undefined) {
    return builder.max(input, builder.constant(options.minValue));
  } else {
    return builder.min(
        builder.max(input, builder.constant(options.minValue)),
        builder.constant(options.maxValue));
  }
}
```

Chromium implementation of [[TFLite] Support other range for Clamp operator](https://issues.chromium.org/issues/326156496) can refer to above sample code for decomposition. So frameworks should stick to spec for consistency.

> > * APIs that allows users to query the implementation status
> 
> This issue [webmachinelearning/webnn#463](https://github.com/webmachinelearning/webnn/issues/463) is talking about exposing operators/types status for each backend.

Correct, https://github.com/webmachinelearning/webnn/issues/463 is the right one for feature detection.

> > * versioning
> 
> I am not aware of discussion about versioning. @huningxin, do you know?

Generally, Web spec doesn't have versioning, WebNN spec follows the same principle. WebNN spec is now in CR status for browser prototyping and developer preview. This kind of interop feedback would be great input to WG. Once the spec moves into more stable status, browser implementation should stick to the latest released version and maintain the backward compatibility.

File: js/web/docs/webnn-operators.md (hunk 1 above)
Comment (Honry) [reply js/web/docs/webnn-operators.md]: Done, thanks!

**Comment by fs-eire** [comment]
@Honry @huningxin thank you for your detailed explanation. 

> https://webmachinelearning.github.io/webnn-status/

This link is helpful. Perhaps I can add this into somewhere in ORT documentation

> how does WebGPU EP manage versioning?

Technically, WebGPU is doing feature detection based on ["capabilities"](https://gpuweb.github.io/gpuweb/#optional-capabilities). For example, the f16 support can be detected by using this API. Good to know that WebNN has this discussion: https://github.com/webmachinelearning/webnn/issues/463.

**Comment by guschmue** [comment]
I think it is clear that the versioning for webnn needs work.
Not sure if comparing webgpu is right since webgpu operates on a lower level and the surface there is more on a feature level that can be used by the ep while webnn has N ops where each op can be subject to changes between versions.

**Review by guschmue — APPROVED** [review]

**Comment by fdwr** [comment]
Wanming, can we implement the [min/max decomposition in ORT](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-clamp) when using the WebNN CPU backend, then remove it when Junwei implements it in Chromium's CPU backend? Then we'd have fewer fragmented partitions, and it would be closer to the expected end result. In any case, I'm glad you added the Chromium todo comment, and I still approve knowing that proper support in the web API is coming soon.

**Review by fdwr — APPROVED** [review]

**Comment by Honry** [comment]
> Wanming, can we implement the [min/max decomposition in ORT](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-clamp) when using the WebNN CPU backend, then remove it when Junwei implements it in Chromium's CPU backend? Then we'd have fewer fragmented partitions, and it would be closer to the expected end result. In any case, I'm glad you added the Chromium todo comment, and I still approve knowing that proper support in the web API is coming soon.

👍Good point, I will follow up.

**Pull request merged**